### PR TITLE
Swap mock & stub implementations

### PR DESCRIPTION
**Pull request by zbuc**

### CODE REVIEW

File: proto/proto/dex.proto
Comment (hdevalence) [inline proto/proto/dex.proto:43]: We should probably note that there is an ordering, though, in order to ensure that there's only one valid `TradingPair` for each pair of assets.

**Comment by hdevalence** [comment]
I think it'd be helpful to pull out the mock flow ciphertexts into a separate PR that we can merge early, and to avoid adding the `Swap`, `SwapClaim` variants to `Transaction` until they're fleshed out.  That way, we can be merging changes to the `Swap`, `SwapClaim` without affecting what data people can post to the chain.

**Comment by zbuc** [comment]
@hdevalence separated the MockFlowCiphertext (and proto types for Swap) into https://github.com/penumbra-zone/penumbra/pull/1055

**Comment by zbuc** [comment]
@hdevalence I think this is at a good merge point now, and the rest of #1047 can be in a different PR

**Comment by hdevalence** [comment]
> I'm not happy with how much I had to put in the `crypto` crate, but to add the `TradingPair` struct to `component` created a circular dependency, I had similar issues dealing with the `Fee` type in `transaction`.

I think this is probably OK. Actually implementing the swaps for real will require writing SNARK circuits that express the swap-related state transitions, and those circuit implementations should live in the `crypto` crate.  So the data types that the proofs are about could be there too.

File: crypto/src/lib.rs
Comment (hdevalence) [inline crypto/src/lib.rs:21]: This could be `pub mod swap;`, without the `pub use` below, if we wanted to isolate the (more niche) swap functionality into one submodule.

File: crypto/src/swap.rs
Comment (hdevalence) [inline crypto/src/swap.rs:21]: Hmm, let's pull the `Fee` out into the `crypto` crate, then.

File: proto/proto/dex.proto
Comment (hdevalence) [inline proto/proto/dex.proto:61]: Yep, we can bundle the constant-length encoding into the encrypt/decrypt functions, since we'll only need to handle it there.

File: proto/proto/dex.proto
Comment (hdevalence) [inline proto/proto/dex.proto:18]: I don't think we need a spend auth signature here, because the `Swap` action consumes the input funds from the transaction's value balance, rather than spending any notes itself.  So the `Swap` will have to be alongside a `Spend` action(s) that unlock the required input funds, and those actions will have the spend authorization signatures for each of those notes already.

File: proto/proto/dex.proto
Comment (hdevalence) [inline proto/proto/dex.proto:36]: This should probably clarify that it's the *block* root, identifying the block the `Swap` was included in.

File: transaction/src/action/swap.rs
Comment (hdevalence) [inline transaction/src/action/swap.rs:13]: We shouldn't use a `SpendProof` here, because the swap proof statement is different.  Instead we should make a new `SwapProof` (it can be a placeholder at the moment), and fill in the swap-specific proof statements there later.

File: transaction/src/transaction.rs
Comment (hdevalence) [inline transaction/src/transaction.rs:81]: just checking, this is blocked on including the actions in the `Transaction` enum right?

Comment (zbuc) [reply transaction/src/transaction.rs:81]: Yes

**Review by hdevalence — APPROVED** [review]
This looks good.  It'd be nice to clean up the warnings before merging, and also to have a `SwapProof` rather than a `SpendProof`, but those could also be done in later PRs.

**Pull request merged**
